src/Components/Web/test/Forms/InputNumberTest.cs

```diff
@@ -1,54 +1,93 @@
     {
         // Arrange
         var model = new TestModel();
-        var rootComponent = new TestInputHostComponent<int, TestInputNumberComponent>
+        var hostComponent = new TestInputHostComponent<int, TestInputNumberComponent>
         {
             EditContext = new EditContext(model),
             ValueExpression = () => model.SomeNumber,
             AdditionalAttributes = new Dictionary<string, object>
-                {
-                    { "DisplayName", "Some number" }
-                }
+            {
+                { "type", "range" }  // User-defined 'type' attribute to override default
+            }
         };
-        var fieldIdentifier = FieldIdentifier.Create(() => model.SomeNumber);
-        var inputComponent = await InputRenderer.RenderAndGetComponent(rootComponent);
 
         // Act
-        await inputComponent.SetCurrentValueAsStringAsync("notANumber");
+        var componentId = await RenderAndGetTestInputNumberComponentIdAsync(hostComponent);
 
-        // Assert
-        var validationMessages = rootComponent.EditContext.GetValidationMessages(fieldIdentifier);
-        Assert.NotEmpty(validationMessages);
-        Assert.Contains("The Some number field must be a number.", validationMessages);
-    }
+        // Retrieve the render tree frames and extract attributes
+        var frames = _testRenderer.GetCurrentRenderTreeFrames(componentId);
 
-    [Fact]
-    public async Task InputElementIsAssignedSuccessfully()
-    {
-        // Arrange
-        var model = new TestModel();
-        var rootComponent = new TestInputHostComponent<int, TestInputNumberComponent>
+        // Debugging: Output the frames for inspection
+        foreach (var frame in frames.Array)
         {
-            EditContext = new EditContext(model),
-            ValueExpression = () => model.SomeNumber,
-        };
+            Console.WriteLine($"Frame: {frame.FrameType}, {frame.ElementName}, {frame.AttributeName}, {frame.AttributeValue}");
+        }
 
-        // Act
-        var inputSelectComponent = await InputRenderer.RenderAndGetComponent(rootComponent);
+        bool inputElementFound = false;
+        var attributes = new Dictionary<string, object>();
+
+        for (int i = 0; i < frames.Count; i++)
+        {
+            var frame = frames.Array[i];
+            if (frame.FrameType == RenderTreeFrameType.Element && frame.ElementName == "input")
+            {
+                inputElementFound = true;
+                for (int j = i + 1; j < frames.Count; j++)
+                {
+                    var attributeFrame = frames.Array[j];
+                    if (attributeFrame.FrameType != RenderTreeFrameType.Attribute)
+                    {
+                        break;
+                    }
+                    attributes[attributeFrame.AttributeName] = attributeFrame.AttributeValue;
+                }
+                break;
+            }
+        }
```

Comment:
I think we should be able to simplify the logic here a bit by utilizing some helpers we've written for other tests.

Something like this:

```csharp
// Act
var hostComponentId = _testRenderer.AssignRootComponentId(hostComponent);
await _testRenderer.RenderRootComponentAsync(hostComponentId);

var batch = _testRenderer.Batches.Single();
var inputComponentId = batch.GetComponentFrames<InputNumber<int>>().Single().ComponentId;
var frames = _testRenderer.GetCurrentRenderTreeFrames(inputComponentId);
var typeAttributeFrame = frames.Array.Single(static frame =>
    frame.FrameType == RenderTreeFrameType.Attribute &&
    frame.AttributeName == "type");

// Assert
Assert.Equal("range", typeAttributeFrame.AttributeValue);
```

And this would let us remove the new helper methods in this class.

Reply:
I have tried to implement the test using the helpers, please let me know if it looks alright. Many thanks !